BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -399,263 +399,329 @@ def browse
 
   # Register a list of files into the system.
   # The files' meta data will be saved as Userfile resources.
-  # This method is (unfortunately) also used to unregister files, and delete them (on the browsable side)
   def register
-    # TODO: refactor completely!
-    @provider = DataProvider.find_accessible_by_user(params[:id], current_user)
+    # Extract key parameters & make sure the provider is browsable
+    @provider   = DataProvider.find_accessible_by_user(params[:id], current_user)
+    @as_user    = browse_as(params['as_user_id'])
 
-    scope      = scope_from_session('data_providers#browse')
-    @as_user   = current_user
-      .available_users
-      .where(:id => scope.custom['as_user_id'] ||= (
-        params['as_user_id'] || current_user.id
-      ))
-      .first
-    @as_user ||= current_user
+    flash[:notice] ||= ''
+    api_request = [:json, :xml].include?(request.format.to_sym)
 
-    unless @provider.is_browsable?(current_user)
-      flash[:error] = "You cannot register files from this provider."
+    ensure_browsable(@provider, 'delete', proc do |html, data, status|
```

Comment:
Is it really worth it to replace a "if" statement containing a respond block (the original code) with a method which takes a three parameter callback ? Repeating the original if statement in all three actions register, unregister and delete would have taken less code in total...